packages/analytics-js/src/components/userSessionManager/UserSessionManager.ts

```diff
@@ -254,14 +268,86 @@ class UserSessionManager implements IUserSessionManager {
+  onError(error: unknown, customMessage?: string): void {
     if (this.errorHandler) {
-      this.errorHandler.onError(error, USER_SESSION_MANAGER);
+      this.errorHandler.onError(error, USER_SESSION_MANAGER, customMessage);
     } else {
       throw error;
     }
   }
 
+  /**
+   * A function to make an external request to set the cookie from server side
+   * @param key       cookie name
+   * @param value     encrypted cookie value
+   */
+  setServerSideCookie(cookieData: CookieData[], store?: IStore): void {
+    try {
+      const encryptedCookieData: EncryptedCookieData[] = [];
+      cookieData.forEach(e => {
+        const encryptedValue =
+          e.value === ''
+            ? e.value
+            : store?.encrypt(stringifyWithoutCircular(e.value, false, [], this.logger));
+        if (isDefinedAndNotNull(encryptedValue)) {
+          encryptedCookieData.push({
+            name: e.name,
+            value: encryptedValue,
+          });
+        }
+      });
+      if (encryptedCookieData.length > 0) {
+        this.httpClient?.getAsyncData({
+          url: `${state.serverCookies.dataServerUrl.value}/rsaRequest`,
+          options: {
+            method: 'POST',
+            data: stringifyWithoutCircular({
+              reqType: 'setCookies',
+              workspaceId: state.source.value?.workspaceId,
+              data: {
+                options: {
+                  maxAge: state.storage.cookie.value?.maxage,
+                  path: state.storage.cookie.value?.path,
+                  domain: state.storage.cookie.value?.domain,
+                  sameSite: state.storage.cookie.value?.samesite,
+                  secure: state.storage.cookie.value?.secure,
+                  expires: state.storage.cookie.value?.expires,
+                },
+                cookies: encryptedCookieData,
+              },
+            }) as string,
+            sendRawData: true,
+          },
+          isRawResponse: true,
+          callback: (res, details) => {
+            if (details?.xhr?.status === 200) {
+              cookieData.forEach(each => {
+                const cookieValue = store?.get(each.name);
+                if (each.value) {
+                  if (cookieValue !== each.value) {
+                    this.logger?.error(FAILED_SETTING_COOKIE_FROM_SERVER_ERROR(each.name));
+                  }
+                } else if (cookieValue) {
+                  this.logger?.error(FAILED_TO_REMOVE_COOKIE_FROM_SERVER_ERROR(each.name));
+                }
+              });
+            } else {
+              this.logger?.error(DATA_SERVER_REQUEST_FAIL_ERROR(details?.xhr?.status));
+              cookieData.forEach(each => {
+                store?.set(each.name, each.value);
+              });
+            }
+          },
+        });
+      }
+    } catch (e) {
+      this.onError(e, FAILED_SETTING_COOKIE_FROM_SERVER_GLOBAL_ERROR);
+      cookieData.forEach(each => {
+        store?.set(each.name, each.value);
```

Comment:
Does it remove the cookie automatically if the value is empty?

Reply:
No, client-side. It's a call to interact with the current store, right?

Reply:
Sounds good. You should also implement this for the other failure scenarios. It's better to pass a callback function from the caller instead of repeating this operation multiple times.

Reply:
What do you mean by "other failure scenarios"?
The callback function and error handling are a little different.

Reply:
If setting the cookie was unsuccessful (but the request got 200), we should still resort to client side cookies. So, implement the fallback mechanism for all such scenarios.
Instead of repeating the code to set or remove the cookies for all these scenarios, you can pass a callback function that handles this.

If you are still need more explanation, let's discuss this offline.

Reply:
Addressed